### DISCUSSION TITLE
File attachments?

### DISCUSSION THREAD

**Discussion by ghost** [Azure/azure-mobile-apps#563] [General]
Does anyone have a good idea on how to associate File attachments to EntityData for offline use in mobile apps? For example I would like to attach pictures/pdf that can also be viewed offline once pulled from the remote table

**Comment by adrianhall** [comment]
This is something I've been thinking about working on.  It's a totally separate product though.

1. Create a FileAttachments table.  Associate each entity in your EntityData to FileAttachments however you wish - it's up to you to maintain this.  Each entity in the FileAttachments table corresponds to a file in Azure blob storage.  This can be provided as a custom repository on the server side that just works with Azure blob storage.
2. Synchronize the FileAttachments table as normal.  (Might be best to have a non-transmitted field in the FileAttachments table that tells you whether the file has been downloaded yet).
3. Hook into the SynchronizationProgress event handler (on the DatasyncClient) - whenever there is a new FileAttachments synchronization event, handle it - for deletions, you can just delete the file; for update/create, you can add the file reference to a queue.  (Ensure the queue is persisted so you can restart downloads when the app starts)
4. Have a separate downloader queue processor (which you kick off when the app starts) that downloads files in the queue to an on-device location.  (Don't forget to update that non-transmitted field in the FileAttachments table)
5. Write an extension method to the IOfflineTable<T> class that basically is "GetFileReferenceAsync(string id)" to get an appropriate file reference for your platform.

It's definitely doable and it's definitely a lot of work.

**Reply by ghost** [reply]
I understand, that explanation makes complete sense.
